flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java

```diff
@@ -76,12 +80,34 @@ public static ClassLoader buildUserCodeClassLoader(
 
 	public static JobExecutionResult submitJob(
 			ClusterClient<?> client,
-			JobGraph jobGraph) throws ProgramInvocationException {
+			JobGraph jobGraph,
+			ClassLoader classLoader) throws ProgramInvocationException {
 		checkNotNull(client);
 		checkNotNull(jobGraph);
 		try {
 			return client
 				.submitJob(jobGraph)
+				.thenApply(jobID -> {
+					// wait till initialization is over. Report errors.
+					JobStatus status;
+					try {
+						status = client.getJobStatus(jobID).get();
+						while (status == JobStatus.INITIALIZING) {
+							Thread.sleep(20L); // TODO which timeout to use?
+							status = client.getJobStatus(jobID).get();
+						}
+						if (status == JobStatus.FAILED) {
+							JobResult result = client.requestJobResult(jobID).get();
+							Optional<SerializedThrowable> maybeError = result.getSerializedThrowable();
+							if (maybeError.isPresent()) {
+								throw maybeError.get().deserializeError(classLoader);
+							}
+						}
+						return jobID;
+					} catch (Throwable e) {
+						throw new CompletionException("Error while waiting for Job initalization", e);
+					}
+				})
```

Comment:
So this blocks needs to be reworked. As general rules,
1) never call `CompletableFuture#get()` within a future, but a return a future instead and use `thenCompose()`,
2) never `Thread.sleep()` within a Future but schedule a new one instead.

You can _pretty much_ use `FutureUtils#retrySuccessfulWithDelay`, except we also want an exponential back-off.

You'd then use it like this:
```
ClassLoader classLoader = ...;

CompletableFuture<Void> submissionCompleteFuture = client
	.submitJob(jobGraph)
	.thenCompose(jobId -> FutureUtils.retrySuccessfulWithDelay(
		() -> client.getJobStatus(jobId),
		Time.of(50, TimeUnit.MILLISECONDS),
		Deadline.fromNow(Duration.ofMinutes(5)),
		jobStatus -> jobStatus != JobStatus.INITIALIZING,
		...)
	.thenApply(status -> Tuple2.of(jobId, status)))
	.thenCompose(jobIdAndStatus -> {
		if (jobIdAndStatus.f1 == JobStatus.FAILED) {
			return client.requestJobResult(jobIdAndStatus.f0)
				.thenAccept(result -> {
					Optional<SerializedThrowable> maybeError = result.getSerializedThrowable();
					if (maybeError.isPresent()) {
						FutureUtils.completedExceptionally(maybeError.get().deserializeError(classLoader));
					}
				});
		} else {
			return CompletableFuture.completedFuture(null);
		}
	});
```

Reply:
Thanks a lot! This is very helpful feedback.

- For some reason, I was working under the assumption that this code is used for the submission through the CliFrontend, however, it is only used for tests. (Should I maybe move these methods into some test util?)

- I agree that exponential backoff is a good idea here. I guess I will extend the `retryOperationWithDelay` with a `WaitStrategy` (relocating these classes), right?

- Getting a ScheduledExecutor here seems tough. 
Creating a `Executors.newScheduledThreadPool(1)` with a `ScheduledExecutorServiceAdapter` seems to be an overkill. ... Which brings me back to building something myself, using the `ForkJoinPool.commonPool()`?

Reply:
1) Well, for about 1 release IIRC they were used for the actual submission. There is already a ticket for moving these, see FLINK-15299.
2) Sounds good.
3) A quick hack would be using `TestingUtils.defaultScheduledExecutor()`, but this would only be available after 1) was finished. A better solution would be to just let users pass in the executor; this isn't unusual and many tests using the `FutureUtills` are doing so already.